File: applications/api-gateway/src/services/project_service.py
# src/services/project_service.py

from sqlalchemy.orm import Session

from src.models.project import Project
from src.models.user import User
from src.schemas.project import ProjectCreate
from src.schemas.project import ProjectUpdate
from src.services.dashboard_service import invalidate_dashboard_cache


def create_project(db: Session, project: ProjectCreate, current_user: User):

    db_project = Project(
        name=project.name,
        platform=project.platform,
        customer=project.customer,
        tenant_id=current_user.tenant_id
    )

    db.add(db_project)
    db.commit()
    db.refresh(db_project)
    invalidate_dashboard_cache(current_user.tenant_id)

    return db_project


def get_projects(db: Session, current_user: User):

    return (
        db.query(Project)
        .filter(Project.tenant_id == current_user.tenant_id)
        .order_by(Project.id.desc())
        .all()
    )


def get_project(db: Session, project_id: int, current_user: User):

    return (
        db.query(Project)
        .filter(Project.id == project_id)
        .filter(Project.tenant_id == current_user.tenant_id)
        .first()
    )


def update_project(db: Session, project_id: int, project: ProjectUpdate, current_user: User):

    db_project = get_project(db, project_id, current_user)

    if db_project is None:
        return None

    db_project.name = project.name
    db_project.platform = project.platform
    db_project.customer = project.customer

    db.commit()
    db.refresh(db_project)
    invalidate_dashboard_cache(current_user.tenant_id)

    return db_project


def delete_project(db: Session, project_id: int, current_user: User):

    db_project = get_project(db, project_id, current_user)

    if db_project is None:
        return None

    db.delete(db_project)
    db.commit()
    invalidate_dashboard_cache(current_user.tenant_id)

    return db_project

File: applications/frontend/src/services/projectService.ts
// src/services/projectService.ts

import { getAccessToken } from "./authService";

const API_URL = import.meta.env.VITE_API_URL ?? "http://localhost:8000";

export type Project = {
  id: number;
  name: string;
  platform: string;
  customer: string;
  tenant_id: number;
};

export type ProjectInput = {
  name: string;
  platform: string;
  customer: string;
};

async function request<T>(url: string, options?: RequestInit): Promise<T> {
  const token = getAccessToken();
  const headers = new Headers(options?.headers);

  if (token) {
    headers.set("Authorization", `Bearer ${token}`);
  }

  const response = await fetch(url, {
    ...options,
    headers,
  });

  if (!response.ok) {
    throw new Error(`Request failed with status ${response.status}`);
  }

  return response.json();
}

export async function getProjects() {
  return request<Project[]>(`${API_URL}/projects`);
}

export async function getProject(projectId: string) {
  return request<Project>(`${API_URL}/projects/${projectId}`);
}

export async function createProject(project: ProjectInput) {
  return request<Project>(`${API_URL}/projects`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(project),
  });
}

export async function updateProject(projectId: number, project: ProjectInput) {
  return request<Project>(`${API_URL}/projects/${projectId}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(project),
  });
}

export async function deleteProject(projectId: number) {
  return request<Project>(`${API_URL}/projects/${projectId}`, {
    method: "DELETE",
  });
}

export type HealthStatus = {
  status: string;
  database: string;
  redis: string;
};

export async function getSystemHealth() {
  return request<HealthStatus>(`${API_URL}/health`);
}

export type DashboardSummary = {
  project_count: number;
  platform_counts: Record<string, number>;
  recent_projects: Project[];
  cached: boolean;
};

export async function getDashboardSummary() {
  return request<DashboardSummary>(`${API_URL}/dashboard/summary`);
}
